### ISSUE TITLE
Indexing fails with 0.27.0 rc versions - MDB_BAD_VALSIZE

### ISSUE THREAD

**Issue by entrptaher** [meilisearch/meilisearch#2338]
**Describe the bug**
Indexing fails with 0.27.0 rc versions when trying to push multiple updates of 5k or 10k elements.

**Logs**
```
   {
      "uid": 4,
      "indexUid": "conversations",
      "status": "failed",
      "type": "documentAddition",
      "details": {
        "receivedDocuments": 5000,
        "indexedDocuments": 0
      },
      "error": {
        "message": "internal: MDB_BAD_VALSIZE: Unsupported size of key/DB name/data, or wrong DUPFIXED size.",
        "code": "internal",
        "type": "internal",
        "link": "https://docs.meilisearch.com/errors#internal"
      },
      "duration": "PT5.847927536S",
      "enqueuedAt": "2022-04-21T08:51:26.066429759Z",
      "startedAt": "2022-04-21T08:51:27.799228957Z",
      "finishedAt": "2022-04-21T08:51:33.647156493Z",
      "batchUid": 4
    },
```

**Meilisearch version:** [v0.27.0], v0.27.0rc0, v0.27.0rc1, v0.27.0rc2

**Comment by Kerollmops** [comment]
Hey @entrptaher,

Thank you for your report. It is an issue that we already had previously, unfortunately, it seems that we changed the tokenizer and that could be the source of the issue. It is surely related to the size of the extracted words from the text, LMDB (our internal key-value store) has a limit on the size of the keys, and keys correspond to the words in the parsed texts.

Would it be possible to send us the dataset you were trying to index, or even better, the part of the dataset that is failing with this error? This way we track the bug down.

Have you tried previous versions of the engine, like v0.26?

Have a nice day 🌤️

**Comment by entrptaher** [comment]
Yes 0.26 worked.

**Comment by entrptaher** [comment]
It’s confidential production data. Maybe I can prepare some dummy data for you.

**Comment by Kerollmops** [comment]
Unfortunately, I was not able to reproduce the bug [by trying to index words that are too long](https://github.com/meilisearch/milli/pull/504), that's the first thing I thought that could produce this bug. I would be glad if you could provide a dummy dataset that is triggering the bug!

**Comment by adicco** [comment]
Perhaps connected to the change of tokenizer, I have noticed, using the Docker image v0.27.0rc2, that when adding documents, already existing ones gets randomly deleted. 

The issues does't happen on v0.26 (current 'latest' image)

Sorry I can't be more precise atm but if this persists I can try to generate a more precise report

**Comment by entrptaher** [comment]
I tried to create a reproducable version of the data from our production data but it's so massive I failed to do so. In the meantime, I tried few more different ways of importing data. Sometimes importing them 50 at once worked a bit longer before throwing the error.

```
[ 
  {
    "uid": 1,
    "indexUid": "conversations",
    "status": "failed",
    "type": "documentAddition",
    "details": {
      "receivedDocuments": 9956,
      "indexedDocuments": 0
    },
    "error": {
      "message": "internal: MDB_BAD_VALSIZE: Unsuppor
ted size of key/DB name/data, or wrong DUPFIXED size.
",
      "code": "internal",
      "type": "internal",
      "link": "https://docs.meilisearch.com/errors#in
ternal"
    },
    "duration": "PT10.724046750S",
    "enqueuedAt": "2022-04-27T06:12:56.32657081Z",
    "startedAt": "2022-04-27T06:12:56.336279011Z",
    "finishedAt": "2022-04-27T06:13:07.060325761Z"
  },
  {
    "uid": 0,
    "indexUid": "conversations",
    "status": "succeeded",
    "type": "documentAddition",
    "details": {
      "receivedDocuments": 10000,
      "indexedDocuments": 10000
    },
    "duration": "PT11.288325756S",
    "enqueuedAt": "2022-04-27T06:12:41.143058806Z",
    "startedAt": "2022-04-27T06:12:41.152575489Z",
    "finishedAt": "2022-04-27T06:12:52.440901245Z"
  }
]

```

**Comment by entrptaher** [comment]
I can see something similar was reported before. https://github.com/meilisearch/MeiliSearch/issues/1677

**Comment by Kerollmops** [comment]
Hey @entrptaher,

Indeed #1677 was triggering the same error message and we found out the issue thanks to the logs, could you please provide them?

> Sometimes importing them 50 at once worked a bit longer before throwing the error.

Would it be possible to have the specific batch that is triggering the error? I mean, are you able to retrieve it? You could probably find it by using the task id and the number of documents you are sending by batch (50 here).

I am pretty sure it is only related to a specific batch and not an accumulation of documents in the database because it was the same for #1677. But maybe I'm wrong.

**Comment by Nextra** [comment]
> I am pretty sure it is only related to a specific batch and not an accumulation of documents in the database because it was the same for https://github.com/meilisearch/meilisearch/issues/1677. But maybe I'm wrong.

I am experiencing the same bug and I can at least confirm that this is not true. I have found the document that is responsible for triggering this bug after manually bisecting 50k documents, and indexing this document in isolation is not an issue. It only breaks once it is in the same index as the other 50k entries.

Since I also have confidentiality concerns, I'm trying to narrow down the conflicting data as much as possible. I hope I can get back to you on that.

-----
You can think of the documents as issue

**Comment by irevoire** [comment]
Thanks a lot, we tried to generate a faulty dataset with our fuzzer but it found nothing, the issue seems too complex 😢

**Comment by Kerollmops** [comment]
Thank you very much @Nextra,

> Since I also have confidentiality concerns, I'm trying to narrow down the conflicting data as much as possible. I hope I can get back to you on that.

I will give you some information if you can compile Meilisearch i.e. `cargo build --release` and change a few lines in the code to help understand where the issue is. The error that is triggered and returned in the task is returned by LMDB, in Rust we use wrapping libraries to call C libraries from Rust "safely", the LMDB wrapping library is called heed.

The error is surely returned from a call to a function that writes into the LMDB database i.e. your index. As you were indexing documents when this happened (you were not updating the setting) it is probably located in [the index_document module](https://github.com/meilisearch/milli/tree/312515dd6b676dcc33e1865e2521370b29c9872b/milli/src/update/index_documents), it can be anywhere but there are more chances for this bug to be located in [these two functions](https://github.com/meilisearch/milli/blob/312515dd6b676dcc33e1865e2521370b29c9872b/milli/src/update/index_documents/helpers/grenad_helpers.rs#L210-L279). I would replace all of the `?` operators by an `.unwrap()`, this way the engine would crash and give you the backtrace.

You probably noticed that I was talking about the milli library and note the Meilisearch engine, it is because the document indexing and search system lay. once you updated the milli library you must change the dependencies of milli from Meilisearch. [Look at **all** of the Cargo.toml of the Meilisearch repository](https://github.com/meilisearch/meilisearch/blob/main/meilisearch-lib/Cargo.toml#L33) and change them to your local path.

```diff
- milli = { git = "https://github.com/meilisearch/milli.git", tag = "v0.26.0" }
+ milli = { path = "the/path/to/my/local/milli" }
```

Once you've done all of this, just compile Meilisearch `cargo build --release` and run it with the backtrace, then send your documents again and wait for the engine to give you a backtrace, it will not crash as actix-web prevent requests from crashing the entire program.

```bash
RUST_BACKTRACE=1 ./target/release/meilisearch
```

Thank you very much for your time, I hope you can help us tackle this bug down.
No worries if you can't ❤️

**Comment by Nextra** [comment]
OK. Here comes the data set: https://gist.github.com/Nextra/3bec5169b7525b6ab4ff807d557bbe32

This is the first ~6000 characters from [the script of the 2001 hit movie SHREK™](https://imsdb.com/scripts/Shrek.html) concatenated with a link that is equivalent in structure to the one that caused my initial problem, with most of it replaced by other topical keywords and random characters.

To trigger the bug, insert this document (just an ID plus the text) 200 times with incrementing IDs. Then, adding a single other variant of this document should fail with `internal: MDB_BAD_VALSIZE: Unsupported size of key/DB name/data, or wrong DUPFIXED size.`

I hope this is of help.

**Comment by curquiza** [comment]
Closed by #2366
